api/include/opentelemetry/baggage/baggage.h

```diff
@@ -18,276 +19,61 @@ namespace baggage
-      {
-        header_s.append(UrlEncode(value.substr(0, metadata_separator)));
-        auto metadata = value.substr(metadata_separator);
-        header_s.append(std::string(metadata.data(), metadata.size()));
-      }
-      else
-      {
-        header_s.append(UrlEncode(value));
-      }
-      return true;
-    });
-    return header_s;
-  }
-
-private:
-  static bool IsPrintableString(nostd::string_view str)
-  {
-    for (const auto ch : str)
-    {
-      if (ch < ' ' || ch > '~')
-      {
-        return false;
-      }
-    }
-
-    return true;
-  }
-
-  static bool IsValidKey(nostd::string_view key) { return key.size() && IsPrintableString(key); }
-
-  static bool IsValidValue(nostd::string_view value) { return IsPrintableString(value); }
-
-  // Uri encode key value pairs before injecting into header
-  // Implementation inspired from : https://golang.org/src/net/url/url.go?s=7851:7884#L264
-  static std::string UrlEncode(nostd::string_view str)
-  {
-    auto to_hex = [](char c) -> char {
-      static const char *hex = "0123456789ABCDEF";
-      return hex[c & 15];
-    };
-
-    std::string ret;
-
-    for (auto c : str)
-    {
-      if (std::isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~')
-      {
-        ret.push_back(c);
-      }
-      else if (c == ' ')
-      {
-        ret.push_back('+');
-      }
-      else
-      {
-        ret.push_back('%');
-        ret.push_back(to_hex(c >> 4));
-        ret.push_back(to_hex(c & 15));
-      }
-    }
-
-    return ret;
-  }
-
-  // Uri decode key value pairs after extracting from header
-  static std::string UrlDecode(nostd::string_view str, bool &err)
-  {
-    auto IsHex = [](char c) {
-      return std::isdigit(c) || (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f');
-    };
-
-    auto from_hex = [](char c) -> char {
-      return std::isdigit(c) ? c - '0' : std::toupper(c) - 'A' + 10;
-    };
-
-    std::string ret;
-
-    for (size_t i = 0; i < str.size(); i++)
-    {
-      if (str[i] == '%')
-      {
-        if (i + 2 >= str.size() || !IsHex(str[i + 1]) || !IsHex(str[i + 2]))
-        {
-          err = 1;
-          return "";
-        }
-        ret.push_back(from_hex(str[i + 1]) << 4 | from_hex(str[i + 2]));
-        i += 2;
-      }
-      else if (str[i] == '+')
-      {
-        ret.push_back(' ');
-      }
-      else if (std::isalnum(str[i]) || str[i] == '-' || str[i] == '_' || str[i] == '.' ||
-               str[i] == '~')
-      {
-        ret.push_back(str[i]);
-      }
-      else
-      {
-        err = 1;
-        return "";
-      }
-    }
-
-    return ret;
-  }
+  std::string ToHeader() const { return baggage_impl_->ToHeader(); }
 
 private:
-  // Store entries in a C-style array to avoid using std::array or std::vector.
-  nostd::unique_ptr<opentelemetry::common::KeyValueProperties> kv_properties_;
+  Baggage(BaggageImpl *baggage_impl) : baggage_impl_(baggage_impl) {}
+  nostd::shared_ptr<BaggageImpl> baggage_impl_;
```

Comment:
Maybe I'm missing something: the `shared_ptr` is wrapped now and not exposed, but isn't it the goal to provide a way to get completely rid of `shared_ptr` usage? How would that work with this solution?

Reply:
You are right @pyohannes. I was additionally trying to make the copy constructor & assignment operator for `Baggage` private, to ensure that we increment ref count only for `BaggageImpl` internally when needed using `friend` class, but seems we can't have non-copyable alternatives in `std::variant`.

Reply:
So, in essence, we are not really making it better-perf for any of the default exporters.. we making it worse by creating yet another abstraction layer with a shared_ptr. Okay. 🤔

Reply:
I'd like to see a Google Benchmark before and after.

Reply:
Yes true. In fact, it adds another layer of redirection, along with a small cost incurred for storing <shared_ptr> as a data member. This should be a solution only if we plan not to use smart pointers across API. It's more of a design decision than a performance solution.

Reply:
Let's run some numbers. I am also interested in synchronous, non-batching, streaming exporter. And in good numbers.

Reply:
> So, in essence, we are not really making it better-perf for any of the default exporters.. we making it worse by creating yet another abstraction layer with a shared_ptr. Okay.

I kind of agree, I don't see how this improves things. We still have a baked-in shared pointer which we hide from the user, and the user has no way to change that.

I don't see any benefits of this refactor. If it's a purely stylistic effort, I have no strong opinion about that.

> Do you see any better approach to handle it?

I see two possible ways:
1. Implementing a purely stack-based solution ourselves. This would require a re-implementation of the underlying key/value data structure.
2. Giving the user means so they can provide their own baggage implementation, which follows their own memory management strategy.

Maybe there are other ways not coming to my mind. If we go with (2), the same approach could also be used to avoid using shared pointers for spans.